Typedefs_Syntax.md

```diff
@@ -0,0 +1,138 @@
+
+# Typedefs Syntax
+
+This document explains the brand new Typedefs Syntax for Typedefs.
+
+## Building blocs
+
+This new Typedefs syntax only has a couple Building blocks: `0`, `1`, `+`, `*` and `^`
+
+- `0` represents the empty type, an irrepresentable type with no inhabitant. Most useful when used
+with `+` and type parameters
+- `1` represents the unit type. It has a single inhabitant which is trivially constructible.
+- `+` Allows you to pick a single value between two possible types.
+- `*` Allows you to package two types together in a single value.
+- `^` Multiple uses of `*`:  `a * a * a = a ^ 3`
+
+For more details look at the [Theoretical tutorial](TUTORIAL_UNDERSTANDING_TYPEDEFS.md)
+
+You can build Typedefs with those simple building blocs by binding them to a name with `:=` and
+ending with `;` :
+
+```
+Bool := 1 + 1;
+
+Byte := Bool ^ 8;
+```
+This example defines a Typedef `Bool` and a Typedef `Byte` which is a product of 8 `Bool`s.
+
+## Type parameters
+
+If the type you want to define takes a type parameter you can define it by following its name by
+variable names:
+
+```
+Maybe a := 1 + a;
+
+Either a b := a + b;
+```
+
+This example defines a Typedef `Maybe` which takes a type parameter `a` and is used as the second
+alterntive to `Maybe`. `Either` takes two type parameters and says that a value is either the first
+one or the second one.
+
+## Type application
+
+If you want to refer to a type that's already been defined you can simply refer to it's name like
+we've seen with the `Byte` example. However, you can only use _fully applied types_.
+
+```
+BoolOrBit := Either Bool Byte;
+```
+
+The example with type parameters can be rewritten this way:
+
+```
+Either a b := a + b;
+
+Maybe a = Either 1 a;
+```
+
+## Recursive types and named sums
+
+If you want to define types that refer to themselves for a tree-like structure, you will need to
+express the type as reference to itself. However this syntax does not work:
+
+```
+List a := 1 + (a * List a)
+```
+
+Indeed we require self-referential types to name each of their constructors using the following
+syntax:
+
+```
+List a := Nil : 1 | Cons : (a * List a)
+```
+
+Where the `|` operator acts like `+` when dealing with named constructors. This syntax is reminicent
+of ML-like languages that allow you to define datatypes with multiple constructors, this is also
+the case here:
+
+```
+Tree a b := Tensor   : (Tree a b) * (Tree a b)
+          | Sequence : (Tree a b) * (Tree a b)
+          | Sym      : a * a
+          | Id       : a
+          | Mor      : b
+```
+
+(An example directly taken from our `FSM-oracle` project)
```

Comment:
```suggestion
(This example was directly taken from our `FSM-oracle` project.)
```
Suggestion: link to the relevant FSM-oracle code?

Reply:
It would be great but the code hasn't been merged yet.